proc/merge/merge.go

```diff
@@ -1,39 +1,274 @@
+// in its own goroutine so that deadlock is avoided when the upstream pullers
+// would otherwise block waiting for an adjacent puller to finish but the
+// Merger is waiting on the upstream puller.
 type Proc struct {
-	parents []proc.Interface
-	merger  *zbuf.Merger
+	ctx  context.Context
+	cmp  expr.CompareFn
+	once sync.Once
+	// parents holds all of the upstream pullers and never changes.
+	parents []*puller
+	// The head-of-line (hol) queue is maintained as a min-heap on cmp of
+	// hol.vals[0] (see Less) so that the next Read always returns
+	// hol[0].vals[0].
+	hol     []*puller
+	blocked map[*puller]struct{}
+	err     error
 }
 
+var _ zbuf.Puller = (*Proc)(nil)
+var _ zio.Reader = (*Proc)(nil)
+
 func New(ctx context.Context, parents []proc.Interface, cmp expr.CompareFn) *Proc {
-	pullers := make([]zbuf.Puller, 0, len(parents))
+	pullers := make([]*puller, 0, len(parents))
 	for _, p := range parents {
-		pullers = append(pullers, p)
+		pullers = append(pullers, newPuller(ctx, p))
 	}
 	return &Proc{
-		parents: parents,
-		merger:  zbuf.NewMerger(ctx, pullers, cmp),
+		ctx:     ctx,
+		cmp:     cmp,
+		parents: pullers,
+		blocked: make(map[*puller]struct{}),
+	}
+}
+
+func (p *Proc) Pull() (zbuf.Batch, error) {
+	p.once.Do(p.run)
+	if p.err != nil {
+		return nil, p.err
+	}
+	if p.Len() == 0 {
+		// No more batches in head of line.  So, lets resume
+		// everything and return an EOS.
+		for _, parent := range p.parents {
+			parent.resumeCh <- struct{}{}
+			ok, err := p.replenish(parent)
+			if err != nil {
+				p.err = err
+				return nil, err
+			}
+			if ok {
+				heap.Push(p, parent)
+				delete(p.blocked, parent)
+			}
+		}
+		return nil, nil
+	}
+	min := heap.Pop(p).(*puller)
+	if p.Len() == 0 || p.cmp(&min.vals[len(min.vals)-1], &p.hol[0].vals[0]) <= 0 {
+		// Either min is the only upstreams or min's last value is less
+		// than or equal to the next upstream's first value.  Either
+		// way, it's safe to return min's remaining values as a batch.
+		batch := min.batch
+		if len(min.vals) < len(batch.Values()) {
+			batch = zbuf.NewArray(min.vals)
+		}
+		ok, err := p.replenish(min)
+		if err != nil {
+			p.err = err
+			return nil, err
+		}
+		if ok {
+			heap.Push(p, min)
+		}
+		return batch, nil
+	}
+	heap.Push(p, min)
+	const batchLen = 100 // XXX
+	vals := make([]zed.Value, 0, batchLen)
+	for len(vals) < batchLen {
+		val, err := p.Read()
+		if err != nil {
+			return nil, err
+		}
+		if val == nil {
+			break
+		}
+		// Copy the underlying buffer because the next call to
+		// zr.Read may overwrite it.
+		vals = append(vals, *val.Copy())
+	}
+	if len(vals) == 0 {
+		return nil, nil
+	}
+	return zbuf.NewArray(vals), nil
+}
+
+func (p *Proc) Read() (*zed.Value, error) {
+	if p.Len() == 0 {
+		return nil, nil
 	}
+	u := p.hol[0]
+	val := &u.vals[0]
+	u.vals = u.vals[1:]
+	if len(u.vals) == 0 {
+		ok, err := p.replenish(u)
+		if err != nil {
+			return nil, err
+		}
+		if !ok {
+			heap.Pop(p)
```

Comment:
My recollection is that these heap operations are spendy and avoiding unnecessary calls to `heap.Fix` in `zbuf.Merger.Read` made a measurable difference.
```suggestion
			heap.Pop(p)
			return val, nil
```